src/tensorlake/function_executor/service.py

```diff
@@ -256,3 +284,165 @@ def get_info(
+                task_execution.updated.notify_all()
+
+    def list_tasks(
+        self, request: ListTasksRequest, context: grpc.ServicerContext
+    ) -> ListTasksResponse:
+        with self._tasks_lock:
+            tasks = []
+            for te in self._running_tasks.values():
+                with te.updated:
+                    tasks.push(te.response)
+
+            return ListTasksResponse(tasks=tasks)
+
+    def start_task(
+        self, request: RunTaskRequest, context: grpc.ServicerContext
+    ) -> StartTaskResponse:
+        self._validate_run_task_request(request)
+
+        with self._tasks_lock:
+            if request.task_id in self._running_tasks:
+                context.abort(
+                    grpc.StatusCode.ALREADY_EXISTS,
+                    f"Task {request.task_id} already exists",
+                )
+
+            task_execution = _TaskExecution(request)
+            self._running_tasks[request.task_id] = task_execution
+
+            task_execution.thread = threading.Thread(
+                target=self._execute_task_in_thread,
+                args=(request, task_execution),
+                daemon=True,
+            )
+            task_execution.thread.start()
+
+        return StartTaskResponse(task_id=request.task_id)
+
+    def await_task(
+        self, request: AwaitTaskRequest, context: grpc.ServicerContext
+    ) -> Generator[TaskStreamResponse, None, None]:
+        """Wait for task completion and stream progress updates."""
+        task_id = request.task_id
+
+        with self._tasks_lock:
+            task_execution = self._running_tasks.get(task_id)
+
+        if task_execution is None:
+            context.abort(
+                grpc.StatusCode.NOT_FOUND,
+                f"Task {task_id} not found",
+            )
+
+        # Stream progress updates until task completes
+        response = None
+        while True:
+            with task_execution.updated:
+                if (
+                    task_execution.response.outcome_code
+                    != TaskOutcomeCode.TASK_OUTCOME_CODE_UNKNOWN
+                ):
+                    response = task_execution.response
+                    break
+
+                task_execution.updated.wait()
+                progress = task_execution.progress
+
+            if progress:
+                yield TaskStreamResponse(progress=progress)
+
+        yield TaskStreamResponse(task_result=response)
+
+    def delete_task(
+        self, request: DeleteTaskRequest, context: grpc.ServicerContext
+    ) -> Empty:
+        """Delete a task and clean up resources."""
+        task_id = request.task_id
+
+        with self._tasks_lock:
+            task_execution = self._running_tasks.get(task_id)
+
+            if task_execution is None:
+                context.abort(
+                    grpc.StatusCode.NOT_FOUND,
+                    f"Task {task_id} not found",
+                )
+
+            with task_execution.updated:
```

Comment:
Shouldn't we just delete the task here if it's finished and return an error otherwise?
Host Executor already deals with these outcomes and failure reasons and it's not easy. Why do we need to do this here?

Reply:
I didn't want to get into a situation where we have an un-deletable task; if the Executor says "Delete this", I think it's the job of a correct FE to make the task go away, and it's not the executor's job to track that the delete failed and that it needs to try again sometime.

Reply:
There're undeletable tasks by design. E.g. this task_execution.updated can block forever because the user function can block forever. Given that I don't understand why all this extra logic and task await is needed here. Also this design forces Executor to use function deadline for this delete_task RPC and extend it if there are progress updates from the function.

Reply:
Done